### PR TITLE
Fix Multiline TextInput with a fixed height scrolls to the bottom when prepending new lines to the text

### PR DESCRIPTION
**Pull request by fabOnReact**
### Please re-state the problem that we are trying to solve in this issue.

Multiline TextInput with a fixed height will scroll to the bottom of the screen when prepending new lines to the text.

### What is the root cause of that problem?

The issue is caused by iOS UITextView:
- The cursor moves to the end of the text when prepending new lines. 
- Moving the cursor to the end of the text triggers the scroll to the bottom.

The behavior was reproduced on an iOS App (without react-native). 
The example included below implements a Component RCTUITextView based on UITextView, which modifies the UITextView attributedText with the textViewDidChange callback (source code available in this [comment](https://github.com/Expensify/App/issues/19507#issuecomment-1595730348)). 

Adding a new line on top of the UITextView on iOS results in:
Issue 1) The cursor moves to the end of TextInput text
Issue 2) The TextInput scrolls to the bottom

<details><summary>Reproducing the issue on an iOS App without react-native</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/246601549-99f480f3-ce80-4678-9378-f71c8aa67e17.mp4" width="900" />

</p>
</details>

Issue 1) is already fixed in react-native, which restores the previous cursor position (on Fabric with  [_setAttributedString](https://github.com/fabriziobertoglio1987/react-native/blob/71e7bbbc2cf21abacf7009e300f5bba737e20d17/packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm#L600-L610)) after changing the text.
Issue 2) needs to be fixed in react-native.

### What changes do you think we should make in order to solve the problem?

Setting the correct TextInput scroll position after re-setting the cursor.

## Changelog:

[IOS] [FIXED] - Fix Multiline TextInput with a fixed height scrolls to the bottom when changing AttributedText

## Test Plan:

Fabric (reproduces on controlled/not controlled TextInput example):

| Before    | After |
| ----------- | ----------- |
| <video src="https://github.com/facebook/react-native/assets/24992535/e06b31fe-407d-4897-b608-73e0cc0f224a" width="350" />      | <video src="https://github.com/facebook/react-native/assets/24992535/fa2eaa31-c616-43c5-9596-f84e7b70d80a" width="350" />       |

Paper (reproduces only on controlled TextInput example):

```javascript
function TextInputExample() {
  const [text, setText] = React.useState('');

  return (
    <View style={{marginTop: 200}}>
      <TextInput
        style={{height: 50, backgroundColor: 'white'}}
        multiline={true}
        value={text}
        onChangeText={text => {
          setText(text);
        }}
      />
    </View>
  );
}
```

| Before    | After |
| ----------- | ----------- |
| <video src="https://github.com/facebook/react-native/assets/24992535/6cb1f2de-717e-4dce-be0a-644f6a051c08" width="350" />      | <video src="https://github.com/facebook/react-native/assets/24992535/dee6edb6-76c6-48b0-b78f-99626235d30e" width="350" />       |

### CODE REVIEW

**Comment by NickGerleman** [comment]
Would this also prevent scroll when adding to the end of the list?

**Comment by fabriziobertoglio1987** [comment]
Thanks for the code review. I moved the PR to draft. I'm working on a solution for this issue.

**Comment by fabriziobertoglio1987** [comment]
https://github.com/facebook/react-native/assets/24992535/14811724-8dc7-40f9-824e-713ff14c45df

**Comment by fabriziobertoglio1987** [comment]
Thanks for the code-review @NickGerleman. I updated the PR and I included a video test in this [comment](https://github.com/facebook/react-native/pull/38679#issuecomment-1684651282).

File: packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm [hunk 1]
```diff
@@ -594,6 +594,22 @@ - (void)_setAttributedString:(NSAttributedString *)attributedString
...
+  BOOL previousScrollEnabled = _backedTextInputView.scrollEnabled;
+  NSInteger previousOffsetStart = [_backedTextInputView offsetFromPosition:_backedTextInputView.beginningOfDocument
+                                                        toPosition:selectedRange.start];
+  if (previousScrollEnabled && previousOffsetStart > 0 && attributedString.string.length >= previousOffsetStart) {
+    NSString *lastChar = [attributedString.string substringWithRange:NSMakeRange(previousOffsetStart - 1, 1)];
+    NSInteger previousOffsetFromEnd = oldTextLength - previousOffsetStart;
+    if (previousOffsetFromEnd > 0 && [lastChar isEqual:@"\n"]) {
+      _backedTextInputView.scrollEnabled = NO;
+    }
+  }
```
Comment (NickGerleman) [inline packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:606]: Could you explain this heuristic? Is it specific to adding newlines?

Comment (fabriziobertoglio1987) [reply packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:606]: Thanks @NickGerleman 

>Could you explain this heuristic?

https://github.com/facebook/react-native/pull/38679/files#diff-65bdfd8888173755da0fd38dae8c70ba1561c11ec453e690195a2f327fa6b62bR603

Scrolling is temporarily disabled when prepending a new line character to the attributed string:
- the cursor is not at the end of the string (previousOffsetFromEnd == 0) 
- the last typed character (lastChar) is a new line character

>Is it specific to adding newlines?

Yes. It is specific to adding new lines.

It is iOS issue [652653](https://developer.apple.com/forums/thread/652653), previously addressed with commit [Maintain cursor position when the text changes in the text input](https://github.com/facebook/react-native/commit/b0e39b2ed9b66b378eb75bee9e692fc801431ddd).

1) Adding a new line changes the text spans and paragraph alignment. The old attributed string may have three spans with different paragraph alignments. The new attributed string could have one span with paragraph alignment 4 (see the test case below for a breakdown). 
2) Changing the AttributedText (text) attributes (alignment) triggers a mutation of the attributed text ([_textOf][1])
3) Mutating the AttributedString causes the cursor to move to the end of the string
4) Moving the cursor to the end of the string causes the TextView to scroll to the end

[1]: https://github.com/facebook/react-native/blob/71e7bbbc2cf21abacf7009e300f5bba737e20d17/packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm#L594 "_textOf"

https://developer.apple.com/forums/thread/652653
https://stackoverflow.com/questions/34914948/how-to-stop-cursor-changing-position-when-i-setattributedtext-in-uitextview-dele

<details><summary>Reproducing the issue on an iOS App without react-native</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/246601549-99f480f3-ce80-4678-9378-f71c8aa67e17.mp4" width="900" />

Sourcecode available at https://github.com/Expensify/App/issues/19507#issuecomment-1607559935.

</p>
</details>


### Details of the changes in the AttributedString attributes (paragraph alignment) when adding new line

**Before** one of the AttributedString spans has NSParagraphStyle Alignment 0. The string text is `"1\n\n2"`.

<details><summary>CLICK TO OPEN ATTRIBUTED STRING DETAILS</summary>
<p>

```
Printing description of oldAttributedString->attributes:
Length 5 (8 blocks, 3 used, block 0 is at 0)
 2 0x600002e70230 {
    NSBackgroundColor = "UIExtendedGrayColorSpace 0 0";
    NSColor = "UIExtendedGrayColorSpace 0 1";
    NSFont = "<UICTFont: 0x130457dc0> font-family: \".SFUI-Regular\"; font-weight: normal; font-style: normal; font-size: 14.00pt";
    NSParagraphStyle = "Alignment 4, LineSpacing 0, ParagraphSpacing 0, ParagraphSpacingBefore 0, HeadIndent 0, TailIndent 0, FirstLineHeadIndent 0, LineHeight 0/0, LineHeightMultiple 0, LineBreakMode 0, Tabs (\n    28L,\n    56L,\n    84L,\n    112L,\n    140L,\n    168L,\n    196L,\n    224L,\n    252L,\n    280L,\n    308L,\n    336L\n), DefaultTabInterval 0, Blocks (\n), Lists (\n), BaseWritingDirection 0, HyphenationFactor 0, TighteningForTruncation NO, HeaderLevel 0 LineBreakStrategy 0 PresentationIntents (\n) ListIntentOrdinal 0 CodeBlockIntentLanguageHint ''";
}
 1 0x600002f9d5e0 {
    NSBackgroundColor = "UIExtendedGrayColorSpace 0 0";
    NSColor = "UIExtendedGrayColorSpace 0 1";
    NSFont = "<UICTFont: 0x130457dc0> font-family: \".SFUI-Regular\"; font-weight: normal; font-style: normal; font-size: 14.00pt";
    NSParagraphStyle = "Alignment 0, LineSpacing 0, ParagraphSpacing 0, ParagraphSpacingBefore 0, HeadIndent 0, TailIndent 0, FirstLineHeadIndent 0, LineHeight 0/0, LineHeightMultiple 0, LineBreakMode 0, Tabs (\n    28L,\n    56L,\n    84L,\n    112L,\n    140L,\n    168L,\n    196L,\n    224L,\n    252L,\n    280L,\n    308L,\n    336L\n), DefaultTabInterval 0, Blocks (\n), Lists (\n), BaseWritingDirection 0, HyphenationFactor 0, TighteningForTruncation NO, HeaderLevel 0 LineBreakStrategy 0 PresentationIntents (\n) ListIntentOrdinal 0 CodeBlockIntentLanguageHint ''";
}
 2 0x600002e70230 {
    NSBackgroundColor = "UIExtendedGrayColorSpace 0 0";
    NSColor = "UIExtendedGrayColorSpace 0 1";
    NSFont = "<UICTFont: 0x130457dc0> font-family: \".SFUI-Regular\"; font-weight: normal; font-style: normal; font-size: 14.00pt";
    NSParagraphStyle = "Alignment 4, LineSpacing 0, ParagraphSpacing 0, ParagraphSpacingBefore 0, HeadIndent 0, TailIndent 0, FirstLineHeadIndent 0, LineHeight 0/0, LineHeightMultiple 0, LineBreakMode 0, Tabs (\n    28L,\n    56L,\n    84L,\n    112L,\n    140L,\n    168L,\n    196L,\n    224L,\n    252L,\n    280L,\n    308L,\n    336L\n), DefaultTabInterval 0, Blocks (\n), Lists (\n), BaseWritingDirection 0, HyphenationFactor 0, TighteningForTruncation NO, HeaderLevel 0 LineBreakStrategy 0 PresentationIntents (\n) ListIntentOrdinal 0 CodeBlockIntentLanguageHint ''";
}
```

</p>
</details>

**After** The AttributedString has only one span. The span has NSParagraphStyle Alignment 4. The string text is `"1\n\n2"`.

<details><summary>CLICK TO OPEN ATTRIBUTED STRING DETAILS</summary>
<p>

```
Printing description of attributedString->attributes:
Length 5 (2 blocks, 1 used, block 0 is at 0)
 5 0x600002e70230 {
    NSBackgroundColor = "UIExtendedGrayColorSpace 0 0";
    NSColor = "UIExtendedGrayColorSpace 0 1";
    NSFont = "<UICTFont: 0x130457dc0> font-family: \".SFUI-Regular\"; font-weight: normal; font-style: normal; font-size: 14.00pt";
    NSParagraphStyle = "Alignment 4, LineSpacing 0, ParagraphSpacing 0, ParagraphSpacingBefore 0, HeadIndent 0, TailIndent 0, FirstLineHeadIndent 0, LineHeight 0/0, LineHeightMultiple 0, LineBreakMode 0, Tabs (\n    28L,\n    56L,\n    84L,\n    112L,\n    140L,\n    168L,\n    196L,\n    224L,\n    252L,\n    280L,\n    308L,\n    336L\n), DefaultTabInterval 0, Blocks (\n), Lists (\n), BaseWritingDirection 0, HyphenationFactor 0, TighteningForTruncation NO, HeaderLevel 0 LineBreakStrategy 0 PresentationIntents (\n) ListIntentOrdinal 0 CodeBlockIntentLanguageHint ''";
}
```

</p>
</details>


I included a video of the two examples at this [link](https://user-images.githubusercontent.com/24992535/250248022-27665b9c-bd23-403f-819e-a9ca438d3350.mp4).

**Comment by fabriziobertoglio1987** [comment]
Thanks @NickGerleman. Could you tell me what is the lint warning? I can fix it and push again the branch so it is easier to import.

**Comment by NickGerleman** [comment]
It's a formatting issue that I can fix before landing, but it is waiting on a reviewer of the imported PR. The best expert for iOS TextInput has been out of office.

**Comment by fabriziobertoglio1987** [comment]
Thanks, @NickGerleman. Maybe the formatting issue is the maximum line length?
I want to push an updated version of this PR to fix the Facebook Internal Linter warning.

**Comment by fabriziobertoglio1987** [comment]
Thanks @NickGerleman. I want to help to get this PR merged.
I have been away for 2+ months, I had to move to a new home, but now I'm again available to help.

**Comment by fabriziobertoglio1987** [comment]
Thanks @NickGerleman. How can I help to get this PR merged?

**Comment by thymikee** [comment]
Looks like this stuck at the Phabricator import. Any chance we can push this forward? :)

**Comment by adhorodyski** [comment]
Hey @NickGerleman, looks like this one's pretty far in the review and just needs some final touches. Is it possible you can jump in to help here?

**Comment by NickGerleman** [comment]
Hey there everyone, sorry for the radio silence. I have worked with text before, and tried to give this a look through, but I don't have much familiarity with the iOS text implementation, to be able to say that this is the right solution, in the right place, etc. I just asked someone with more familiarity there to take a look. But otherwise, will take a look at shipping this next week.

File: packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm [hunk 2]
```diff
@@ -593,6 +593,22 @@ - (void)_setAttributedString:(NSAttributedString *)attributedString
   }
```
[the remark this answers is not in the document]
Comment (fabOnReact) [reply packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm]: https://github.com/facebook/react-native/pull/38679#discussion_r1487054891
>I'm for fixing UIKit problems and making it better in React Native.

Yes. It is a UIKit problem (https://github.com/Expensify/App/issues/19507#issuecomment-1595730348,  https://github.com/Expensify/App/issues/19507#issuecomment-1607559935 and https://developer.apple.com/forums/thread/652653).

>If we wanted to consider a matrix of different scroll offsets, characters, cursor positions, etc, what would the full test matrix look like?

I prepared the following test cases:

<details><summary>Sourcecode of the Example</summary>
<p>

```jsx
function TextInputExample() {
  const [text, setText] = React.useState('');
  const [color, setColor] = React.useState('black');

  const changeColorAndText = () => {
    if (color === 'black') {
      setColor('red');
    } else {
      setColor('black');
    }
    setText(text + ' updated.');
  };

  const changeColor = () => {
    if (color === 'black') {
      setColor('red');
    } else {
      setColor('black');
    }
  };

  const appendText = () => {
    setText(text + ' updated.');
  };

  return (
    <View style={{marginTop: 200}}>
      <TextInput
        multiline={true}
        style={[styles.text]}
        onChangeText={text => setText(text)}>
        <Text style={{color: color}}>{text}</Text>
      </TextInput>
      <Button title="change color" onPress={changeColor} />
      <Button title="change color and text" onPress={changeColorAndText} />
      <Button title="append text" onPress={appendText} />
    </View>
  );
}
```

</p>
</details>

- Text selection is not empty ([after][1])
- Adding a new paragraph (\n) at position 0, then deleting it ([after][3])
- Adding a new paragraph (\n) in the middle of the text ([after][5])
- Changing the text color ([after][7])
- Adding text to the end of the string ([after][9])
- Changing the text characters ([after][11])
- Increasing the line height ([after][12])

[1]: https://drive.google.com/file/d/1Pt08TNMJwmzoLzbLvSITrbKIy2KxMVEo/view?usp=sharing
[3]: https://drive.google.com/file/d/12rzpbwi4QZrDQhp18Ra2WUygr1oVm8fr/view?usp=share_link
[5]:  https://drive.google.com/file/d/1NgorHR2Qr-a8Fo5fcghqvQhWBIWBn49K/view?usp=share_link
[7]: https://drive.google.com/file/d/1AdnHawXRVx7Z0exVRlsGH3c8i4Kx5Hpl/view?usp=share_link
[9]: https://drive.google.com/file/d/1ACUcOd7kCLV-UA2KYDvWIHynjbITngZG/view?usp=share_link
[11]: https://drive.google.com/file/d/1X569mepp-Sse3oDE3aDfL2rDQhFxZoMJ/view?usp=share_link
[12]: https://drive.google.com/file/d/1SKEjC5Bc7u6ip_YvsJBNFdhiil7nW79L/view?usp=share_link

File: packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm [hunk 3]
```diff
@@ -593,6 +593,22 @@ - (void)_setAttributedString:(NSAttributedString *)attributedString
...
   UITextRange *selectedRange = _backedTextInputView.selectedTextRange;
   NSInteger oldTextLength = _backedTextInputView.attributedText.string.length;
+  BOOL previousScrollEnabled = _backedTextInputView.scrollEnabled;
+  NSInteger previousOffsetStart = [_backedTextInputView offsetFromPosition:_backedTextInputView.beginningOfDocument
+                                                        toPosition:selectedRange.start];
+  if (previousScrollEnabled && previousOffsetStart > 0 && attributedString.string.length >= previousOffsetStart) {
+    NSString *lastChar = [attributedString.string substringWithRange:NSMakeRange(previousOffsetStart - 1, 1)];
+    NSInteger previousOffsetFromEnd = oldTextLength - previousOffsetStart;
+    if (previousOffsetFromEnd > 0 && [lastChar isEqual:@"\n"]) {
```
[the remark this answers is not in the document]
Comment (fabOnReact) [reply packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm]: https://github.com/facebook/react-native/pull/38679#discussion_r1487019009
>this is oddly specific. If anything else besides new line is inserted, it is fine?

I updated the PR to handle any change in the AttributedString (for example color, adding or deleting a paragraph, changing the lineHeight). https://github.com/Expensify/App/issues/19507#issuecomment-1607559935

>Could be good to keep an explanation of how newline specifically effects paragraph composition, which is why it is being special cased.

**FIRST EXAMPLE:** Adding a new paragraph with the new line character `\n` changes the UITextView AttributedText and moves the cursor to the end of the text (https://github.com/Expensify/App/issues/19507#issuecomment-1615584137).

**SECOND EXAMPLE:** Modifying Text styles lineHeight or color causes mutation of the UITextView AttributedText and moves the cursor to the end of the text (https://github.com/Expensify/App/issues/19507#issuecomment-1607559935).

1) The Text minimumLineHeight changes from 0 to 20 
2) The UITextView attributes lineHeight changes from 0 to 20 (screenshot)
3) The change in the AttributesText causes the TextInput cursor to move to the end of the text. The change in position of the cursor triggers scroll position to the end of the text.

UIKit issue https://developer.apple.com/forums/thread/652653.

**Comment by fabOnReact** [comment]
Thanks @NickGerleman. I replied to your code-review and further improved the solution as by Meta requirement.
https://github.com/facebook/react-native/pull/38679#discussion_r1493960840
https://github.com/facebook/react-native/pull/38679#discussion_r1493960647

**Comment by NickGerleman** [comment]
> Thanks @NickGerleman. I replied to your code-review and further improved the solution as by Meta requirement. [#38679 (comment)](https://github.com/facebook/react-native/pull/38679#discussion_r1493960840) [#38679 (comment)](https://github.com/facebook/react-native/pull/38679#discussion_r1493960647)

I will take another look at this soon (still need to write the accompanying internal tests first).

**Comment by NickGerleman** [comment]
> > Thanks @NickGerleman. I replied to your code-review and further improved the solution as by Meta requirement. [#38679 (comment)](https://github.com/facebook/react-native/pull/38679#discussion_r1493960840) [#38679 (comment)](https://github.com/facebook/react-native/pull/38679#discussion_r1493960647)
> 
> I will take another look at this soon (still need to write the accompanying internal tests first).

Finally cleared some other commitments, so writing those tests now.

**Pull request closed without merging**